File: app/src/main/java/com/github/ykiselev/lwjgl3/fs/ResourceFolder.java
package com.github.ykiselev.lwjgl3.fs;

import java.net.URL;
import java.util.Optional;
import java.util.stream.Stream;

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
public interface ResourceFolder {

    default Optional<URL> resolve(String resource) {
        return resolve(resource, true);
    }

    Optional<URL> resolve(String resource, boolean shouldExist);

    Stream<URL> resolveAll(String resource);

    default boolean isWritable() {
        return false;
    }
}

File: app/src/main/kotlin/com/github/ykiselev/playground/services/fs/ResourceFolder.kt
package com.github.ykiselev.playground.services.fs

import java.net.URL

/**
 * @author Yuriy Kiselev (uze@yandex.ru).
 */
interface ResourceFolder {

    fun resolve(resource: String): URL? {
        return resolve(resource, true)
    }

    fun resolve(resource: String, shouldExist: Boolean): URL?
    fun resolveAll(resource: String): Sequence<URL>
    val isWritable: Boolean
        get() = false
}